File: book/petrol-pump/chapter3/section6/Fill.java
package chapter3.section6;

import pump.*;
import chapter3.section5.AccumulatePulsesPump;
import sodium.*;
import java.util.Optional;

public class Fill {
    public final Behavior<Double> price;
    public final Behavior<Double> dollarsDelivered;
    public final Behavior<Double> litersDelivered;

    public Fill(
            Event<Unit> eClearAccumulator, Event<Integer> eFuelPulses,
            Behavior<Double> calibration, Behavior<Double> price1,
            Behavior<Double> price2, Behavior<Double> price3,
            Event<Fuel> eStart) {
        price = capturePrice(eStart, price1, price2, price3);
        litersDelivered = AccumulatePulsesPump.accumulate(
                eClearAccumulator, eFuelPulses, calibration);
        dollarsDelivered = Behavior.lift(
                (liters, price_) -> liters * price_,
                litersDelivered, price);
    }

    public static Behavior<Double> capturePrice(
            Event<Fuel> eStart,
            Behavior<Double> price1, Behavior<Double> price2,
            Behavior<Double> price3) {
        Event<Optional<Double>> ePrice1 = eStart.snapshot(price1,
            (f, p) -> f == Fuel.ONE ? Optional.of(p)
                                    : Optional.empty());
        Event<Optional<Double>> ePrice2 = eStart.snapshot(price2,
            (f, p) -> f == Fuel.TWO ? Optional.of(p)
                                    : Optional.empty());
        Event<Optional<Double>> ePrice3 = eStart.snapshot(price3,
            (f, p) -> f == Fuel.THREE ? Optional.of(p)
                                      : Optional.empty());

        return Event.filterOptional(
            ePrice1.merge(ePrice2.merge(ePrice3))
        ).hold(0.0);
    }
}

File: c#/Book/PetrolPump/PetrolPump/PetrolPump/Chapter3/Section6/Fill.cs
using System;

using PetrolPump.Chapter3.Section5;
using PetrolPump.Pump;

using Sodium;

namespace PetrolPump.Chapter3.Section6
{
public class Fill {
    public readonly Behavior<Double> Price;
    public readonly Behavior<Double> DollarsDelivered;
    public readonly Behavior<Double> LitersDelivered;

    public Fill(
            Event<Unit> eClearAccumulator, Event<int> eFuelPulses,
            Behavior<Double> calibration, Behavior<Double> price1,
            Behavior<Double> price2, Behavior<Double> price3,
            Event<Fuel> eStart) 
    {
        Price = CapturePrice(eStart, price1, price2, price3);
        LitersDelivered = AccumulatePulsesPump.Accumulate(
                eClearAccumulator, eFuelPulses, calibration);
        DollarsDelivered = Behavior<double>.Lift(
                (liters, price_) => liters * price_,
                LitersDelivered, Price);
    }

    public static Behavior<Double> CapturePrice(
            Event<Fuel> eStart,
            Behavior<Double> price1, Behavior<Double> price2,
            Behavior<Double> price3) 
    {
        Event<Optional<Double>> ePrice1 = eStart.Snapshot(price1,
            (f, p) => f == Fuel.ONE ? Optional<double>.Of(p)
                                    : Optional<double>.Empty());
        Event<Optional<Double>> ePrice2 = eStart.Snapshot(price2,
            (f, p) => f == Fuel.TWO ? Optional<double>.Of(p)
                                    : Optional<double>.Empty());
        Event<Optional<Double>> ePrice3 = eStart.Snapshot(price3,
            (f, p) => f == Fuel.THREE ? Optional<double>.Of(p)
                                      : Optional<double>.Empty());

        return Event<double>.FilterOptional(
            ePrice1.Merge(ePrice2.Merge(ePrice3))
        ).Hold(0.0);
    }
  }
}
